### PR TITLE
[FIX] web_editor: change header style of mega menu

### PR DESCRIPTION
**Pull request by deso-odoo**
Current behavior before PR:

- Some website snippets contained elements with classes like h1, h2, h3, etc. `setTag` function did not remove these classes when changing the tag, resulting in no visual effect.
- In commit [[1]](https://github.com/odoo-dev/odoo/commit/b4cf3bcc893e345c25d863c0afaab71bbbfb404c), when fixing the `setTagName` function to avoid adding a `<p>` tag when converting a tag to a `<p>` inside an `<li>`, it missed a scenario where the element was deeply nested within the `<li>` tag. For instance, in a case like `<li><div><h1>abcd</h1></div></li>`, the function failed to replace the `<h1>`with a `<p>` tag.
- Commit [[2]](https://github.com/odoo-dev/odoo/commit/d04e32c6f9da2b2e8709985648786df6f7eb6091) introduces an approach to preserve the cursor in `setTag` when new node is inserted. It used `setStart` at `firstLeaf` of `startContainer` and used offset of the `startContainer` similarly for `endContainer` which is incorrect and would throw traceback regarding no child at that offset.
- `destroyLinkTools` function sets the selection to entire link. However, in case where a website snippet had a structure like
```html
<a>
    <div>
        <i class=fa-xxx>​</i>
        <div>
            <h4>Text</h4>
            <font>Text</font>
        </div>
    </div>
</a>
```
selecting the complete link caused problem. The toolbar couldn't be updated correctly, also one could not change the a tag of a single element within the link.

Desired behavior after PR is merged:

- Classes like h1, h2, h3, etc. are now removed when changing the tag.
- The `setTagName` function has been fixed to correctly replace a tag with a `<p>` tag when deeply nested inside an `<li>`.
- Fixed it by getting the correct `startContainer` and `endContainer` when new node is created.
- `destroyLinkTools` selects the `anchorNode` and the `focusnode` of the selection instead of entire link.

task-3245819

### CODE REVIEW

**Comment by deso-odoo** [comment]
This PR is continuation of https://github.com/odoo/odoo/pull/117304

**Comment by deso-odoo** [comment]
@Zynton As per your request in #117304, I have moved the commit's which we intend to proceed with in this PR. Additionally, as QSM mentioned that the tours for `linkTools.js` are currently disabled on runbot. I've found another file for tour which is `addons/website/static/tests/tours/edit_megamenu.js`.
Regarding the third commit I have changed the code as I realized that just changing from `range.setStart` and `range.setEnd` to `setSelection` was incorrect :eyes:.

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 1]
```diff
@@ -35,6 +35,7 @@ const preserveCursor = OdooEditorLib.preserveCursor;
 const closestElement = OdooEditorLib.closestElement;
 const setSelection = OdooEditorLib.setSelection;
 const endPos = OdooEditorLib.endPos;
+const getCursorDirection = OdooEditorLib.getCursorDirection;
```
Comment (Zynton) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:38]: Commit message:

```diff
@@ -2,14 +2,24 @@ Before this commit:
 
 `destroyLinkTools` function sets the selection to entire link. However, in
 case where a website snippet had a structure like
-`<a><div><i class=fa-xxx>​</i><div><h4>Text</h4><font>Text</font></div></div>
-</a>` selecting the complete link caused problem. The toolbar couldn't be
+```html
+<a>
+    <div>
+        <i class=fa-xxx>\u200b​</i>
+        <div>
+            <h4>Text</h4>
+            <font>Text</font>
+        </div>
+    </div>
+</a>
+```
+selecting the complete link caused problem. The toolbar couldn't be
 updated correctly, also one could not change the a tag of a single element
 within the link.
 
 After this commit:
 
-Fix destroyLinkTools such that now it select the anchorNode and the focus
-node of selection instead of complete link.
+`destroyLinkTools` selects the `anchorNode` and the `focusNode` of the
+selection instead of the entire link.
 
 task-3245819
```

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 2]
```diff
@@ -1261,8 +1262,17 @@ const Wysiwyg = Widget.extend({
...
                     }
                 } else {
-                    anchorNode = link;
-                    focusNode = link;
+                    const selectionDirection = getCursorDirection(selection.anchorNode, 0, selection.focusNode, 0);
+                    if (
+                        closestElement(selection.anchorNode, 'a') === link &&
+                        closestElement(selection.focusNode, 'a') === link
+                    ) {
+                        [anchorNode, focusNode] = selectionDirection
```
Comment (Zynton) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:1270]: I know just testing `selectionDirection` is truthy is enough because `DIRECTIONS.RIGHT === true` but I'd still prefer if it were tested explicitly because that's really not intuitive when you read it.

```suggestion
                    const isDirectionRight = getCursorDirection(selection.anchorNode, 0, selection.focusNode, 0) === DIRECTIONS.RIGHT;
                    if (
                        closestElement(selection.anchorNode, 'a') === link &&
                        closestElement(selection.focusNode, 'a') === link
                    ) {
                        [anchorNode, focusNode] = isDirectionRight

File: addons/website/static/tests/tours/edit_megamenu.js [hunk 3]
```diff
@@ -92,4 +92,68 @@ tour.register('edit_megamenu', {
...
+    },
+    {
+        content: "Convert it to Bold",
+        trigger: '#oe_snippets #toolbar #bold',
+    },
+    ...wTourUtils.clickOnSave(),
+    toggleMegaMenu(),
+    {
+        content: "The menu item should only convert selected text to Heading 1.",
+        trigger: '.s_mega_menu_big_icons_subtitles .row > div:first-child .nav > :first-child .media-body > :last-child:not(:has(strong))',
```
Comment (Zynton) [inline addons/website/static/tests/tours/edit_megamenu.js:155]: @qsm-odoo Does it seem ok to you to test with a specific template and its quite specific html content? Seems like something that's gonna be in the way for every little redesign and at the same time it's a tour so why not.

File: addons/web_editor/static/lib/odoo-editor/src/commands/commands.js [hunk 4]
```diff
@@ -379,10 +381,9 @@ export const editorCommands = {
                 children.forEach(child => newBlock.appendChild(child));
             }
         }
-        const newRange = new Range();
-        newRange.setStart(startContainer,startOffset);
-        newRange.setEnd(endContainer,endOffset);
-        getDeepRange(editor.editable, { range: newRange, select: true, });
+        startContainer = !startContainer.isConnected || ['TD', 'DIV', 'LI'].includes(startContainer.nodeName) ? startContainerChild.parentNode : startContainer;
+        endContainer = !endContainer.isConnected || ['TD', 'DIV', 'LI'].includes(endContainer.nodeName) ? endContainerChild.parentNode : endContainer;
+        setSelection(startContainer, startOffset, endContainer, endOffset);
```
[the remark this answers is not in the document]
Comment (deso-odoo) [reply addons/web_editor/static/lib/odoo-editor/src/commands/commands.js:386]: When converting from one block tag to another after shift+enter for eg: `<p>abc<br>[]</p>` here the startContainer would be the `p` tag so converting it to `setTag('h1')` the `p` tag would no longer be present it dom.

[the remark this answers is not in the document]
Comment (Zynton) [reply addons/web_editor/static/lib/odoo-editor/src/commands/commands.js]: Since we call `setTag` just above (in the `for` loop) and that's where we lose the reference, it could be helped by updating the reference there. But if we're sure there is no risk of disconnecting the container children (I checked and if I'm not mistaken it's currently the case) I guess it's ok. This is clearer albeit potentially less robust, but we'll get an error if it's not connected so if a change ever makes disconnecting the container children possible, we should be warned.

tldr: ok with this change.

File: addons/website/static/tests/tours/edit_megamenu.js (hunk 3 above)
Comment (deso-odoo) [reply addons/website/static/tests/tours/edit_megamenu.js]: @qsm-odoo Could you please have a look at the tour.

Comment (qsm-odoo) [reply addons/website/static/tests/tours/edit_megamenu.js]: @Zynton I did not see your ping sorry. A sub-comment in a review to which I am already indirectly subscribed as a code-owner, I'd say it's close to a 50% chance that I see it :grimacing: In general, you can do a ping in the main review comment, I'll understand I'll have to scroll through the sub comments... :pray: (Which I often do anyway to be sure but like I said 50% chance to catch it in that case, if I don't want my job to be just about "scrolling through emails" ^^).

For the test, I think it's ok to rely on the current template, it's not like we change mega menu design every 3 weeks. However, are you sure this should not be about making a lower-level test about "supporting links with complex structure inside"? In that case, would be easier to make a test that will not have to change at long-term? But as you wish :+1:

File: addons/website/static/tests/tours/edit_megamenu.js [hunk 5]
```diff
@@ -92,4 +92,68 @@ tour.register('edit_megamenu', {
...
+        run: 'click',
+    },
+    {
+        content: "Save the changes to the menu",
+        trigger: '.modal-dialog .btn-primary span:contains("Save")',
+        run: 'click',
+    },
+    // Edit a menu item
+    wTourUtils.clickOnEdit(),
+    wTourUtils.clickOnExtraMenuItem({extra_trigger: '#oe_snippets.o_loaded'}),
```
Comment (Zynton) [inline addons/website/static/tests/tours/edit_megamenu.js:131]: Could you add unit tests to cover these flows in a more targeted way?

Comment (deso-odoo) [reply addons/website/static/tests/tours/edit_megamenu.js:131]: The reason for not writing the unit test in the first place was, that the fix is in `destroyLinkTools` and it is triggered when clicking on the side toolbar. I couldn't find a way to access the side toolbar on website or in mass_mailing's QUnit's, I could only find tours when side toolbar are used.

Comment (Zynton) [reply addons/website/static/tests/tours/edit_megamenu.js:131]: If it's specific to the sidebar then indeed this is probably your best bet.

File: addons/web_editor/static/lib/odoo-editor/src/commands/commands.js [hunk 6]
```diff
@@ -356,7 +356,9 @@ export const editorCommands = {
         const deepestSelectedBlocks = selectedBlocks.filter(block => (
             !descendants(block).some(descendant => selectedBlocks.includes(descendant))
         ));
-        const [startContainer, startOffset, endContainer, endOffset] = [firstLeaf(range.startContainer), range.startOffset, lastLeaf(range.endContainer), range.endOffset];
+        let { startContainer, startOffset, endContainer, endOffset } = range;
+        const startContainerChild = startContainer.firstChild;
+        const endContainerChild = endContainer.firstChild;
```
[the remark this answers is not in the document]
Comment (deso-odoo) [reply addons/web_editor/static/lib/odoo-editor/src/commands/commands.js]: We use the child node solely for referencing its parent, so using either `firstChild` or `lastChild` will not break anything. The `endContainerChild` node itself is not utilized; we only need its parent, which remains the same for both `firstChild` and `lastChild`. We could even keep it as `endContainer.firstChild` if you want me to.

File: addons/web_editor/static/lib/odoo-editor/src/commands/commands.js [hunk 7]
```diff
@@ -379,10 +379,14 @@ export const editorCommands = {
                 children.forEach(child => newBlock.appendChild(child));
             }
         }
-        const newRange = new Range();
-        newRange.setStart(startContainer,startOffset);
-        newRange.setEnd(endContainer,endOffset);
-        getDeepRange(editor.editable, { range: newRange, select: true, });
```
Comment (Zynton) [inline addons/web_editor/static/lib/odoo-editor/src/commands/commands.js:384]: Don't we want to keep the deep range where we can? Why not keep this logic after your `if`s instead of using `setSelection`?

File: addons/web_editor/static/lib/odoo-editor/src/commands/commands.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (Zynton) [reply addons/web_editor/static/lib/odoo-editor/src/commands/commands.js]: Ah indeed I took another look and now understand better what happened. I left a new comment as a result. Thanks for bearing with me.

File: addons/web_editor/static/lib/odoo-editor/src/commands/commands.js (hunk 7 above)
Comment (deso-odoo) [reply addons/web_editor/static/lib/odoo-editor/src/commands/commands.js:384]: Indeed we should keep the deep range, I have made the changes...

File: addons/web_editor/static/src/js/editor/snippets.editor.js [hunk 8]
```diff
@@ -1489,7 +1489,7 @@ var SnippetsMenu = Widget.extend({
         // on text changes
         const alreadySelectedElements = new Set();
         this.$document.on('click.snippets_menu', '.o_default_snippet_text', function (ev) {
-            const el = ev.currentTarget;
+            const el = ev.target;
```
Comment (deso-odoo) [inline addons/web_editor/static/src/js/editor/snippets.editor.js:1492]: https://github.com/odoo/odoo/pull/145925#discussion_r1458830385 continued

@Zynton The runbot failed due to rebasing. A recent PR https://github.com/odoo/odoo/pull/144121 [commit](https://github.com/odoo/odoo/pull/144121/commits/c6ce2e7a70c107f1622b71c870b02ff721c4da82) changed the behaviour of clicks on `o_default_snippet_text` to select the complete node when clicked. However, It uses `ev.currentTarget` as element. But for our mega menu case with snippet like
```html
<a class="o_default_snippet_text">
    <div>
        <i class=fa-xxx>​</i>
        <div>
            <h4>Text</h4>
            <font>Text</font>
        </div>
    </div>
</a>
```
When clicking on `h4` element, `ev.currentTarget` returns the `a` tag, as the property doesn't change during event propogation. regardless of where the event is originated. It would always refer to the element to which the event handler is attached, so in this case the `el` variable is set to the `a` tag rather than `h4` tag, causing the entire `a` tag to be selected when clicking on `h4`. So I have changed the el to `ev.target` instead, which represents the actual element on which the event occured. Would you agree with this change?? cc. @qsm-odoo

Comment (qsm-odoo) [reply addons/web_editor/static/src/js/editor/snippets.editor.js:1492]: Hello :wave: 

- I indeed changed `.target` to `.currentTarget` intentionally: "on click on a node marked as default text, select its content" being more accurate than "on click inside a node marked as default, select the inner node that was clicked". Especially since nothing guarantees that that "inner clicked" node is actually something the user actually sees: it could be a simple span without style for all we know :shrug: I probably should have mentioned it in the commit message. In any case, changing it back to `.target` does not seem to be the way to go (at least I don't understand your need).

- With my commit, I actually introduced new bugs that I still have to solve (unrelated to currentTarget/target) but the `o_default_snippet_text` is not removed properly anymore when it should after typing inside nodes, I still have to check that. That bug was reported by RDE here: https://github.com/odoo/odoo/pull/144121#discussion_r1438344278 but then I could not reproduce it, then he could not reproduce it anymore either... then after full forward-port, I found out there actually is the bug he described :shrug: Don't know how or why yet.

Comment (Zynton) [reply addons/web_editor/static/src/js/editor/snippets.editor.js:1492]: I'm in agreement with @qsm-odoo here. What I'm wondering is (1) why it's not satisfactory to select the whole `<a>` as that seems like precisely the feature's goal; and (2) why that breaks your test.

Comment (deso-odoo) [reply addons/web_editor/static/src/js/editor/snippets.editor.js]: I intended to apply bold to `h4` in the tour. However, clicking on the `h4`, the selection moved to `a` tag, causing my tour to fail, Nevertheless, I have adapted the tour, it should be good now.

File: addons/website/static/tests/tours/edit_megamenu.js [hunk 9]
```diff
@@ -92,4 +92,76 @@ tour.register('edit_megamenu', {
...
+    {
+        content: "Select the h4 of first menu link of the first column",
+        trigger: '.s_mega_menu_big_icons_subtitles .row > div:first-child .nav > :first-child .media-body > h4',
+        run: function (actions) {
+            actions.click();
+            const range = document.createRange();
+            range.selectNodeContents(this.$anchor[0]);
+            const sel = window.getSelection();
+            sel.removeAllRanges();
+            sel.addRange(range);
```
Comment (Zynton) [inline addons/website/static/tests/tours/edit_megamenu.js:152]: Please add a comment so it's clear why you had to do that.

Comment (deso-odoo) [reply addons/website/static/tests/tours/edit_megamenu.js:152]: How about
```
            // Clicking on the h4 element first time leads to the selection of
            // the entire a.nav-link, due to presence of `o_default_snippet_text` class
            // hence, specify the selection on the h4 element
```

Comment (Zynton) [reply addons/website/static/tests/tours/edit_megamenu.js:152]: Let's go :-)

Comment (deso-odoo) [reply addons/website/static/tests/tours/edit_megamenu.js:155]: Runbot green!! Good to go... :)

**Comment by Zynton** [comment]
It was an epic journey but we got there, nice job :+1:

@robodoo rebase-ff r+

**Comment by deso-odoo** [comment]
@robodoo retry

File: addons/website/static/tests/tours/edit_megamenu.js [hunk 10]
```diff
@@ -92,4 +92,80 @@ tour.register('edit_megamenu', {
...
+        content: "Click on a menu item",
+        trigger: '#top_menu .nav-item a',
+    },
+    {
+        content: "Click on 'Link' to open Link Dialog",
+        trigger: '.o_edit_menu_popover a.js_edit_menu',
+    },
+    {
+        content: "Trigger the link dialog (click 'Add Mega Menu Item')",
+        extra_trigger: ".o_web_editor_dialog:visible",
```
Comment (rdeodoo) [inline addons/website/static/tests/tours/edit_megamenu.js:110]: No need to change since it's r+ but fyi the trigger are always checked against visible elements, `:visible` is useless

File: addons/website/static/tests/tours/edit_megamenu.js [hunk 11]
```diff
@@ -92,4 +92,80 @@ tour.register('edit_megamenu', {
...
+        trigger: '.modal-body a.js_add_menu[data-type="mega"]',
+    },
+    {
+        content: "Write a label for the new menu item",
+        extra_trigger: ".o_link_dialog",
```
Comment (rdeodoo) [inline addons/website/static/tests/tours/edit_megamenu.js:115]: This is useless as the trigger already contains that part

File: addons/website/static/tests/tours/edit_megamenu.js [hunk 12]
```diff
@@ -92,4 +92,80 @@ tour.register('edit_megamenu', {
...
+        trigger: '.o_link_dialog #o_link_dialog_label_input',
+        run: 'text Megaaaaa!',
+    },
+    {
+        content: "Save the new menu item",
+        trigger: '.modal-dialog .btn-primary span:contains("Save")',
+        run: 'click',
```
Comment (rdeodoo) [inline addons/website/static/tests/tours/edit_megamenu.js:122]: This is useless as default behavior

File: addons/website/static/tests/tours/edit_megamenu.js [hunk 13]
```diff
@@ -92,4 +92,80 @@ tour.register('edit_megamenu', {
...
+    },
+    {
+        content: "Write a label for the new menu item",
+        extra_trigger: ".o_link_dialog",
+        trigger: '.o_link_dialog #o_link_dialog_label_input',
+        run: 'text Megaaaaa!',
+    },
+    {
+        content: "Save the new menu item",
+        trigger: '.modal-dialog .btn-primary span:contains("Save")',
```
Comment (rdeodoo) [inline addons/website/static/tests/tours/edit_megamenu.js:121]: Since there is only one primary btn, you can generally omit the "save" part which doesn't bring anything but will fail in versions where we change the save term (from experience this happens)

File: addons/website/static/tests/tours/edit_megamenu.js [hunk 14]
```diff
@@ -92,4 +92,80 @@ tour.register('edit_megamenu', {
...
+        run: 'click',
+    },
+    {
+        content: "Save the changes to the menu",
+        trigger: '.modal-dialog .btn-primary span:contains("Save")',
+        run: 'click',
```
Comment (rdeodoo) [inline addons/website/static/tests/tours/edit_megamenu.js:127]: useless click

File: addons/website/static/tests/tours/edit_megamenu.js [hunk 15]
```diff
@@ -92,4 +92,80 @@ tour.register('edit_megamenu', {
...
+    },
+    // Edit a menu item
+    wTourUtils.clickOnEdit(),
```
Comment (rdeodoo) [inline addons/website/static/tests/tours/edit_megamenu.js:130]: This is probably a source of race condition, since you click on edit right after having click on "save", but you actually don't await the modal save page reload, meaning you will have 2 concurrents rpc: it could click on "edit" before the rpc was back/processed
There should be an "check step" which check for the new menu in the menu (meaning the page has fully reload)

**Review by rdeodoo — COMMENTED** [review]
Maybe worth a r- to check the "race condition" I mentioned (might be wrong but pretty sure it's not)
Edit: It's actually not in merge mode (r+) anymore, saw it afterwards

File: addons/website/static/tests/tours/edit_megamenu.js (hunk 13 above)
Comment (deso-odoo) [reply addons/website/static/tests/tours/edit_megamenu.js:121]: There are two modal-dialog open here so there are multiple `.btn-primary`
image

Comment (rdeodoo) [reply addons/website/static/tests/tours/edit_megamenu.js:121]: And FYI even if there is multiple modal opened, adding save wouldn't change anything because they both have save as btn ;-)

**Comment by deso-odoo** [comment]
@rdeodoo Is it ok now?

File: addons/website/static/tests/tours/edit_megamenu.js [hunk 16]
```diff
@@ -92,4 +92,83 @@ tour.register('edit_megamenu', {
...
+    {
+        content: "Save the changes to the menu",
+        trigger: '.modal-footer .btn-primary',
+    },
+    wTourUtils.clickOnExtraMenuItem(),
+    {
+        content: "Check for the new mega menu",
+        trigger: '#top_menu .nav-item a.o_mega_menu_toggle:contains("Megaaaaa2!")',
+        run: function () {}, // it's a check
+    },
```
Comment (rdeodoo) [inline addons/website/static/tests/tours/edit_megamenu.js:131]: Followup of https://github.com/odoo/odoo/pull/145925/files#r1464527216
I had a deeper look, right now the modal is actually not saved when you click on it: it calls for a page reload while remaining open, meaning elements behind it won't be considered with normal trigger (they are considered hidden).
So there was probably no race condition possible anyway.. but only while we keep this "keep open" behavior. So it's still best to consider that the popup could be closed on "Save" click and prevent race condition, instead of not considering it and making it break the day we change that (easy win before it even happens).

> rdeodoo Is it ok now?

So the answer is not really: those steps will still be "true" directly after clicking save (before the page is actually reloaded, and considering the case of the modal closing on click).
It's generally how you can find race condition:
```js
{
    content: "Something",
    trigger: '.element1',
}, {
    content: "Something else",
    trigger: '.element2,
},
```
If you do that, you are not waiting that the behavior which happened after clicking on element1 is finished before clicking on element2 -> Race condition everyday.

It's what is being done here (well, not really, since the modal doesn't close on save, but you got it):
```js
{
    content: "Save the changes to the menu",
    trigger: '.modal-footer .btn-primary',
},
wTourUtils.clickOnExtraMenuItem(),
{
    content: "Check for the new mega menu",
    trigger: '#top_menu .nav-item a.o_mega_menu_toggle:contains("Megaaaaa2!")',
    run: function () {}, // it's a check
},
```
- You click on save, which fires a rpc and then a page reload when the rpc is back
- Right after having click on save, you click on extra menu item, without waiting the rpc/page reload, that trigger is immediately true/found

With the current state of things, it is exactly doing the same as before: still the same room for race condition should we make the click on save close the modal. But since it's not closing it, it's ignoring the mega menu item "hidden" behind the modal, and it's actually waiting the page reload "by chance" (because we don't close the modal).

If we were to hide the modal on click (and then wait for the rpc to reload), it would actually click on extra menu item, then wait for the "Megaaaa2" items which doesn't exist since the page is not yet reloaded. Then the page would reload, and the tour would be stuck (because Megaaa2 would be in the DOM now but the mega menu would not be opened, as the mega menu opening step was wrongly done before the page reload).

At the end, I guess it should should be about

```suggestion
    {
        content: "Save the changes to the menu",
        trigger: '.modal-footer .btn-primary',
    },
    {
        content: "Check for the new mega menu",
        trigger: '#top_menu:has(.nav-item a.o_mega_menu_toggle:contains("Megaaaaa2!"))',
        run: function () {}, // it's a check
    },
    wTourUtils.clickOnExtraMenuItem(),
```
This `has()` should really wait correctly the page reload, and will workaround the fact it's a hidden element which we want to be used in a trigger step (as #top_menu is visible, it will work).

To be confirmed, just wrote that on github.

Note: since I figured after having a deeper look that there is no race condition possible thanks to the modal not closing itself on "save" click, you can ignore this if you prefer. I still wanted to take time to explain you how those race conditions works.

Comment (deso-odoo) [reply addons/website/static/tests/tours/edit_megamenu.js]: Thanks for the explanation!!! :)

**Comment by deso-odoo** [comment]
> It was an epic journey but we got there,

@Zynton Not quite, we are still one step away; remember https://github.com/odoo/odoo/pull/117304#discussion_r1422278968
